### PR TITLE
add cmake doc

### PR DESCRIPTION
**Pull request by hegner**
BEGINRELEASENOTES
- Add documentation for CMake macros
ENDRELEASENOTES

### CODE REVIEW

**Comment by hegner** [comment]
This addresses #604

File: doc/cmake.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+PODIO exposes multiple CMake macros to enable the building of data model libraries.
+For a package developer tutorial on CMake, please have a look [here])https://hsf-training.github.io/hsf-training-cmake-webpage/). PODIO follows the same conventions used therein.
+
+## Generate the C++ code
+After a proper `find_package(PODIO)`, the C++ code of the new data model `NewDM` can be created and compiled via
+
+```
+# generate the c++ code from the yaml definition
+PODIO_GENERATE_DATAMODEL(newdm newdm.yaml headers sources IO_BACKEND_HANDLERS)
```
Comment (tmadlener) [inline doc/cmake.md:11]:
```suggestion
PODIO_GENERATE_DATAMODEL(newdm newdm.yaml headers sources)
```

`IO_BACKEND_HANDLERS` needs an argument.

File: doc/cmake.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+# compile the core data model shared library (no I/O)
+PODIO_ADD_DATAMODEL_CORE_LIB(newdm "${headers}" "${sources}")
+```
+
+However, this does not create any I/O components yet. For this, one needs to add the `IO_BACKEND_HANDLERS` parameter and then generate and compile the corresponding backends, like shown below for the ROOT and SIO backends:
```
Comment (tmadlener) [inline doc/cmake.md:17]: Actually by default we always build the `ROOT` backend (even if no `IO_BACKEND_HANDLERS` are passed explictly).

File: doc/cmake.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
+# Building a new data model library with CMake
+
+PODIO exposes multiple CMake macros to enable the building of data model libraries.
+For a package developer tutorial on CMake, please have a look [here])https://hsf-training.github.io/hsf-training-cmake-webpage/). PODIO follows the same conventions used therein.
+
+## Generate the C++ code
+After a proper `find_package(PODIO)`, the C++ code of the new data model `NewDM` can be created and compiled via
+
+```
```
Comment (tmadlener) [inline doc/cmake.md:9]:
```suggestion
```cmake
```

Doesn't really change that much, but it gives a bit of syntax highlighting for cmake.

**Review by tmadlener — COMMENTED** [review]
I think this should also mention the additional arguments that are available from `PODIO_GENERATE_DATAMODEL`. At least the `UPSTREAM_EDM` (with a pointer to [EDM4eic](https://github.com/eic/EDM4eic/blob/main/CMakeLists.txt) potentially) and the `LANG` arguments.

The `DEPENDS` one is at least touched upon at the datamodel syntax introduction.

I also think this has to be added to `doc/index.rst` at an appropriate place.

**Comment by m-fila** [comment]
What do you think about adding a link to macro definitions in  [podioMacros.cmake](https://github.com/AIDASoft/podio/blob/master/cmake/podioMacros.cmake) for reference? In other projects sometimes they put their user-facing cmake macros in some obscure places and I wish they simply put a link in docs to save me from greping everything :sweat_smile:

File: doc/cmake.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+PODIO exposes multiple CMake macros to enable the building of data model libraries.
+For a package developer tutorial on CMake, please have a look [here])https://hsf-training.github.io/hsf-training-cmake-webpage/). PODIO follows the same conventions used therein.
+
+## Exported CMake functions and macros
+The file [cmake/podioMacros.cmake](https://github.com/AIDASoft/podio/blob/master/cmake/podioMacros.cmake) provides the following functions for external use:
+
+  1. PODIO_GENERATE_DATAMODEL - generate the data model from provided yaml file
+  2. PODIO_ADD_DATAMODEL_CORE_LIB - compile the generated data model
+  3. PODIO_ADD_ROOT_IO_DICT - configure ROOT I/O backend and interface to Python
+  4. PODIO_ADD_SIO_IO_BLOCKS - configure SIO backend
```
Comment (tmadlener) [inline doc/cmake.md:12]:
```suggestion
  1. `PODIO_GENERATE_DATAMODEL` - generate the data model from provided yaml file
  2. `PODIO_ADD_DATAMODEL_CORE_LIB` - compile the generated data model
  3. `PODIO_ADD_ROOT_IO_DICT` - configure ROOT I/O backend and interface to Python
  4. `PODIO_ADD_SIO_IO_BLOCKS` - configure SIO backend
```

Highlight them slightly.

File: doc/cmake.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+Their exact parameters and functionality are documented in the file mentioned above. Below a few examples.
+
+## A simple example
+After a proper `find_package(PODIO)`, the C++ code of the new data model `NewDM` can be created and compiled via
```
Comment (tmadlener) [inline doc/cmake.md:17]:
```suggestion
After a proper `find_package(PODIO)`, the C++ code of the new data model `newdm` can be created and compiled via
```
Just for consistency with the snippets below (or alternatively change the ones below).

File: doc/cmake.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+PODIO_GENERATE_DATAMODEL(newdm newdm.yaml headers sources IO_BACKEND_HANDLERS "ROOT;SIO")
+
+# compile the core data model shared library (no I/O)
+PODIO_ADD_DATAMODEL_CORE_LIB(newdm "${headers}" "${sources}")
+
+# generate and compile the ROOT I/O dictionary for the default I/O system
+PODIO_ADD_ROOT_IO_DICT(newdmDict newmdm "${headers}" src/selection.xml)
+
+# compile the SIOBlocks shared library for the SIO backend
+PODIO_ADD_SIO_IO_BLOCKS(edm4hep "${headers}" "${sources}")
```
Comment (tmadlener) [inline doc/cmake.md:42]:
```suggestion
PODIO_ADD_SIO_IO_BLOCKS(newdm "${headers}" "${sources}")
```
Minor typo.

File: doc/cmake.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
+# Building a new data model library with CMake
+
+PODIO exposes multiple CMake macros to enable the building of data model libraries.
+For a package developer tutorial on CMake, please have a look [here])https://hsf-training.github.io/hsf-training-cmake-webpage/). PODIO follows the same conventions used therein.
```
Comment (tmadlener) [inline doc/cmake.md:4]:
```suggestion
For a package developer tutorial on CMake, please have a look [here](https://hsf-training.github.io/hsf-training-cmake-webpage/). PODIO follows the same conventions used therein.
```

To render the link correctly.

File: doc/cmake.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+  1. `PODIO_GENERATE_DATAMODEL` - generate the data model from provided yaml file
+  2. `PODIO_ADD_DATAMODEL_CORE_LIB` - compile the generated data model
+  3. `PODIO_ADD_ROOT_IO_DICT` - configure ROOT I/O backend and interface to Python
+  4. `PODIO_ADD_SIO_IO_BLOCKS` - configure SIO backend
+
+Their exact parameters and functionality are documented in the file mentioned above. Below a few examples.
+
+## A simple example
+After a proper `find_package(PODIO)`, the C++ code of the new data model `newdm` can be created and compiled via
```
Comment (andresailer) [inline doc/cmake.md:17]:
```suggestion
After a proper `find_package(podio REQUIRED)`, the C++ code of the new data model `newdm` can be created and compiled via
```

**Comment by tmadlener** [comment]
Should we merge this now and add the advanced usage in another PR?

**Pull request merged**